src/c.cpp

```diff
@@ -0,0 +1,358 @@
+{
+    opentracing::FinishSpanOptions cppOptions;
+    cppOptions.finish_steady_timestamp =
+        timespecToTimePoint<opentracing::SteadyClock>(
+            options->finish_timestamp);
+    span->impl_->FinishWithOptions(cppOptions);
+    delete span->impl_;
+    delete span;
+}
+
+extern "C" void opentracing_finish_span(opentracing_span_t* span)
+{
+    span->impl_->Finish();
+    delete span->impl_;
+    delete span;
+}
+
+extern "C" void opentracing_set_operation_name(
+    opentracing_span_t* span, const char* operation_name)
+{
+    span->impl_->SetOperationName(operation_name);
+}
+
+extern "C" void opentracing_set_tag(
+    opentracing_span_t* span,
+    const char* key,
+    const opentracing_value_t* value)
+{
+    span->impl_->SetTag(key, value->impl_);
+}
+
+extern "C" void opentracing_set_baggage_item(
+    opentracing_span_t* span,
+    const char* key,
+    const char* value)
+{
+    span->impl_->SetBaggageItem(key, value);
+}
+
+extern "C" const char* opentracing_baggage_item(
+    opentracing_span_t* span,
+    const char* key)
+{
+    return span->impl_->BaggageItem(key).c_str();
+}
+
+/* TODO
+extern "C" void opentracing_log(
+    opentracing_span_t* span,
+    const opentracing_tag_t fields[])
+{
+    span->impl_->Log(fields);
+}*/
+
+extern "C" void opentracing_context_from_span(
+    opentracing_spancontext_t* span_context,
+    const opentracing_span_t* span)
+{
+    span_context->impl_ = const_cast<opentracing::SpanContext*>(
+        &span->impl_->context());
+}
+
+extern "C" void opentracing_tracer_from_span(
+    opentracing_tracer_t* tracer,
+    const opentracing_span_t* span)
+{
+    tracer->impl_ = const_cast<opentracing::Tracer*>(&span->impl_->tracer());
+}
+
+extern "C" opentracing_span_t* opentracing_start_span(
+    opentracing_tracer_t* tracer,
+    const char* operation_name)
+{
+    std::unique_ptr<opentracing_span_t> span(new opentracing_span_t());
+    span->impl_ = tracer->impl_->StartSpan(operation_name).release();
+    return span.release();
+}
+
+extern "C" opentracing_span_t* opentracing_start_span_with_options(
+    opentracing_tracer_t* tracer,
+    const char* operation_name,
+    const opentracing_startspanoptions_t* options)
+{
+    opentracing::StartSpanOptions cppOptions;
+    cppOptions.start_steady_timestamp =
+        timespecToTimePoint<opentracing::SteadyClock>(
+            options->start_steady_timestamp);
+    cppOptions.start_system_timestamp =
+        timespecToTimePoint<opentracing::SystemClock>(
+            options->start_system_timestamp);
+
+    for (auto i = 0; i < options->num_references; ++i) {
+        const auto* ref = options->references[i];
+        cppOptions.references.emplace_back(
+                static_cast<opentracing::SpanReferenceType>(ref->type),
+                ref->context->impl_);
+    }
+
+    for (auto i = 0; i < options->num_tags; ++i) {
+        const auto* tag = options->tags[i];
+        cppOptions.tags.emplace_back(tag->key, tag->value->impl_);
```

Comment:
Functions like this can throw `std::bad_alloc` which wouldn't be caught anywhere.

Reply:
You are right. I'll wrap everything in try blocks to avoid that.